### PR TITLE
Add support to RunServer for notifying clients via condition variable that the interop server has started.

### PR DESCRIPTION
**Pull request by dwittmer**
Change adds a version of grpc::testing::interop::RunServer that allows clients to pass in optional condition_variable which will be notified when the server has started and port to use, avoiding the need for callers to work with command line options.

Above is used to support clients running the server in a separate thread in the same process as the tests are run to support local only tests.

Existing grpc::testing::interop::RunServer calls into new version passing in FLAGS_port and null condition_variable.

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* One or more commits was unable to be associated with a GitHub user. The email used in the Git commit must match a verified email address in a GitHub profile. For more information, please see https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/

_Regards,
The Linux Foundation CLA GitHub bot_

**Comment by dwittmer** [comment]
<rerun bots>

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @dwittmer isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
The Linux Foundation CLA GitHub bot_

**Comment by dwittmer** [comment]
Signed CLA as an employee.

File: test/cpp/interop/interop_server.cc [hunk 1]
```diff
@@ -333,6 +340,10 @@ void grpc::testing::interop::RunServer(
   }
   std::unique_ptr<Server> server(builder.BuildAndStart());
   gpr_log(GPR_INFO, "Server listening on %s", server_address.str().c_str());
+
+  // Signal that the server has started.
+  if (server_started_condition) server_started_condition->notify_all();
```
Comment (dgquintas) [inline test/cpp/interop/interop_server.cc]: where's the wait on the condvar?

Comment (dwittmer) [reply test/cpp/interop/interop_server.cc]: The wait on the condvar would be in the calling code. This allows a client to start the server in a new / separate thread and then wait on the condvar until the server was started.

File: test/cpp/interop/server_helper.h [hunk 2]
```diff
@@ -50,8 +51,22 @@ class InteropServerContextInspector {
...
+/// Run gRPC interop server using port FLAGS_port.
+///
+/// \param creds The credentials associated with the server.
 void RunServer(std::shared_ptr<ServerCredentials> creds);
 
+/// Run gRPC interop server.
+///
+/// \param creds The credentials associated with the server.
+/// \param port Port to use for the server.
+/// \param server_started_condition (optional) Condition variable to used to
```
Comment (dgquintas) [inline test/cpp/interop/server_helper.h]: typo: s/to used/to be used

File: test/cpp/interop/interop_server.cc (hunk 1 above)
Comment (yang-g) [reply test/cpp/interop/interop_server.cc]: The client may get falsely notified from a spurious wakeup. Not sure whether that matters for you though.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: test/cpp/interop/interop_server.cc (hunk 1 above)
Comment (dwittmer) [reply test/cpp/interop/interop_server.cc]: Thanks for bringing this up! Added code to allow clients to correctly detect & handle a spurious wakeup.

Comment (yang-g) [reply test/cpp/interop/interop_server.cc]: LGTM

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yang-g** [comment]
Not sure whether you can see the link but you need to clang-format as follows:
+ tools/dockerfile/grpc_clang_format/clang_format_all_the_things.sh
--- /var/local/git/grpc/test/cpp/interop/server_helper.h	2017-11-17 21:50:57.058366108 +0000
+++ /tmp/tmp.ew6jgE4g9M	2017-11-17 21:51:37.478366108 +0000
@@ -69,9 +69,8 @@
 /// \param port Port to use for the server.
 /// \param server_started_condition (optional) Struct holding mutex, condition
 ///     variable, and condition used to notify when the server has started.
-void RunServer(std::shared_ptr<ServerCredentials> creds,
-               int port,
-               ServerStartedCondition *server_started_condition);
+void RunServer(std::shared_ptr<ServerCredentials> creds, int port,
+               ServerStartedCondition* server_started_condition);
 
 }  // namespace interop
 }  // namespace testing
--- /var/local/git/grpc/test/cpp/interop/interop_server.cc	2017-11-17 21:50:57.058366108 +0000
+++ /tmp/tmp.zQ8spwh0Cj	2017-11-17 21:51:38.942366108 +0000
@@ -321,9 +321,8 @@
 }
 
 void grpc::testing::interop::RunServer(
-    std::shared_ptr<ServerCredentials> creds,
-    const int port,
-    ServerStartedCondition *server_started_condition) {
+    std::shared_ptr<ServerCredentials> creds, const int port,
+    ServerStartedCondition* server_started_condition) {
   GPR_ASSERT(port != 0);
   std::ostringstream server_address;
   server_address << "0.0.0.0:" << port;

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yang-g** [comment]
Test flakes: 
C++ ASAN/TSAN #13122 
Basic tests C++: #13626

**Pull request merged**